theseus/optimizer/nonlinear/nonlinear_optimizer.py

```diff
@@ -122,56 +135,48 @@ def _init_info(
             err_history=err_history,
         )
 
-        # Only copy best solution if needed (None means track_best_solution=False)
-
     def _update_info(
         self,
         info: NonlinearOptimizerInfo,
         current_iter: int,
-        best_err: Optional[torch.Tensor],
         err: torch.Tensor,
         converged_indices: torch.Tensor,
-    ) -> torch.Tensor:
+    ):
         info.converged_iter += 1 - converged_indices.long()
         if info.err_history is not None:
             assert err.grad_fn is None
             info.err_history[:, current_iter + 1] = err.clone().cpu()
-        if info.best_solution is None:
-            return best_err
-        # Only copy best solution if needed (None means track_best_solution=False)
-        assert best_err is not None
-        good_indices = err < best_err
-        info.best_iter[good_indices] = current_iter
-        for var in self.linear_solver.linearization.ordering:
-            info.best_solution[var.name][good_indices] = (
-                var.data.detach().clone()[good_indices].cpu()
-            )
-        return torch.minimum(best_err, err)
 
-    # `track_best_solution` keeps a **detached** copy (as in no gradient info)
-    # of the best variables found, but it is optional to avoid unnecessary copying
-    # if this is not needed
-    #
-    # if verbose, info will also keep track of the full error history
-    def _optimize_impl(
-        self,
-        track_best_solution: bool = False,
-        verbose: bool = False,
-        **kwargs,
-    ) -> OptimizerInfo:
-        # All errors are only used for stopping conditions, so they are outside
-        # compute graph
-        last_err = self.objective.error_squared_norm().detach() / 2
+        if info.best_solution is not None:
+            # Only copy best solution if needed (None means track_best_solution=False)
+            assert info.best_err is not None
+            good_indices = err < info.best_err
+            info.best_iter[good_indices] = current_iter
+            for var in self.linear_solver.linearization.ordering:
+                info.best_solution[var.name][good_indices] = (
+                    var.data.detach().clone()[good_indices].cpu()
+                )
 
-        if verbose:
-            print(
-                f"Nonlinear optimizer. Iteration: {0}. Error: {last_err.mean().item()}"
-            )
+            info.best_err = torch.minimum(info.best_err, err)
 
-        best_err = last_err.clone() if track_best_solution else None
-        converged_indices = torch.zeros_like(last_err).bool()
-        info = self._init_info(last_err, track_best_solution, verbose)
-        for it_ in range(self.params.max_iterations):
+        converged_indices = self._check_convergence(err, info.last_err)
+        info.status[
+            np.array(converged_indices.detach().cpu())
+        ] = NonlinearOptimizerStatus.CONVERGED
+
+    # loop for the iterative optimizer
+    def _optimize_loop(
+        self,
```

Comment:
Missing type hints and also `track_best_solution` is not used.

Reply:
Added type hints and removed `track_best_solution` here